src/cpp/fastdds/domain/DomainParticipantImpl.cpp

```diff
@@ -1843,6 +1857,367 @@ ReturnCode_t DomainParticipantImpl::unregister_type(
+    rpc::ServiceImpl* service(nullptr);
+    
+    try
+    {
+        service = new rpc::ServiceImpl(service_name, service_type_name, this);  
+    }
+    catch (const std::exception& e)
+    {
+        EPROSIMA_LOG_ERROR(PARTICIPANT, "Error creating service: " << e.what());
+        return nullptr;
+    }
+
+    if (service->is_valid())
+    {
+        std::lock_guard<std::mutex> lock(mtx_services_);
+        services_[service_name] = service;
+    }
+    else
+    {
+        delete service;
+        service = nullptr;
+    }
+
+    return service;
+}
+
+rpc::Service* DomainParticipantImpl::find_service(
+        const std::string& service_name) const
+{
+    std::lock_guard<std::mutex> lock(mtx_services_);
+    auto it = services_.find(service_name);
+
+    if (it != services_.end())
+    {
+        return it->second;
+    }
+
+    return nullptr;
+}
+
+ReturnCode_t DomainParticipantImpl::delete_service(
+        const rpc::Service* service)
+{
+    if (!service)
+    {
+        return RETCODE_BAD_PARAMETER;
+    }
+
+    const rpc::ServiceImpl* service_impl = dynamic_cast<const rpc::ServiceImpl*>(service);
+    assert(service_impl != nullptr);
+
+    // Check that service belongs to this participant
+    if (service_impl->participant_ != this)
+    {
+        EPROSIMA_LOG_ERROR(PARTICIPANT, "Service does not belong to this participant.");
+        return RETCODE_PRECONDITION_NOT_MET;
+    }
+
+    std::lock_guard<std::mutex> lock(mtx_services_);
+    auto it = services_.find(service->get_service_name());
+
+    if (it != services_.end())
+    {
+        delete it->second;
+        services_.erase(it);
+        return RETCODE_OK;
+    }
+
+    // The service was not found in this participant
+    return RETCODE_PRECONDITION_NOT_MET;
+}
+
+rpc::Requester* DomainParticipantImpl::create_service_requester(
+        rpc::Service* service,
+        const RequesterQos& qos)
+{
+    // Check if the service is valid and registered in participant
+    if (!service)
+    {
+        EPROSIMA_LOG_ERROR(PARTICIPANT, "Service is nullptr.");
+        return nullptr;
+    }
+
+    const rpc::ServiceImpl* service_impl = dynamic_cast<const rpc::ServiceImpl*>(service);
+    assert(service_impl != nullptr);
+
+    // Check that service belongs to this participant
+    if (service_impl->participant_ != this)
+    {
+        EPROSIMA_LOG_ERROR(PARTICIPANT, "Service does not belong to this participant.");
+        return nullptr;
+    }
+
+    auto it = services_.find(service->get_service_name());
+    if (it == services_.end())
+    {
+        EPROSIMA_LOG_ERROR(PARTICIPANT, "Service with name '" << service->get_service_name() << "' not found.");
+        return nullptr;
+    }
+
+    // Make sure that both services are using the same service type
+    if (it->second->get_service_type_name() != service->get_service_type_name())
+    {
+        EPROSIMA_LOG_ERROR(PARTICIPANT, "Service type mismatch.");
+        return nullptr;
+    }
```

Comment:
I prefer checking that the pointers are the same:
```suggestion
    if (it->second != service_impl)
    {
        EPROSIMA_LOG_ERROR(PARTICIPANT, "Service mismatch.");
        return nullptr;
    }
```

Reply:
Changes applied on f6b7659